### PR TITLE
Fix `stride(A, i)` for 0-dim inputs

### LINKED ISSUE #44087: Regression: stride errors on 0-dim array
Julia 1.7.2:
```julia
julia> stride( fill(1), 1)
1
```

#44027 modified `stride`.

master-branch
```julia
julia> stride( fill(1) , 1)
ERROR: ArgumentError: reducing with add_sum over an empty collection of element type Union{} is not allowed.
You may be able to prevent this error by supplying an `init` value to the reducer.
Stacktrace:
  [1] _empty_reduce_error(f::Any, T::Type)
    @ Base ./reduce.jl:307
  [2] reduce_empty(#unused#::typeof(Base.add_sum), #unused#::Core.TypeofBottom)
    @ Base ./reduce.jl:338
  [3] reduce_empty(op::Base.BottomRF{typeof(Base.add_sum)}, #unused#::Type{Union{}})
    @ Base ./reduce.jl:347
  [4] reduce_empty_iter
    @ ./reduce.jl:371 [inlined]
  [5] reduce_empty_iter
    @ ./reduce.jl:370 [inlined]
  [6] foldl_impl(op::Base.BottomRF{typeof(Base.add_sum)}, nt::Base._InitialValue, itr::Tuple{})
    @ Base ./reduce.jl:49
  [7] mapfoldl_impl(f::typeof(identity), op::typeof(Base.add_sum), nt::Base._InitialValue, itr::Tuple{})
    @ Base ./reduce.jl:44
  [8] mapfoldl(f::Function, op::Function, itr::Tuple{}; init::Base._InitialValue)
    @ Base ./reduce.jl:162
  [9] mapfoldl
    @ ./reduce.jl:162 [inlined]
 [10] #mapreduce#264
    @ ./reduce.jl:294 [inlined]
 [11] mapreduce(f::Function, op::Function, itr::Tuple{})
    @ Base ./reduce.jl:294
 [12] sum(f::Function, a::Tuple{}; kw::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ Base ./reduce.jl:520
 [13] sum(f::Function, a::Tuple{})
    @ Base ./reduce.jl:520
 [14] sum(a::Tuple{}; kw::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ Base ./reduce.jl:549
 [15] sum(a::Tuple{})
    @ Base ./reduce.jl:549
 [16] stride(A::Array{Int64, 0}, k::Int64)
    @ Main ./REPL[7]:4
 [17] top-level scope
    @ REPL[9]:1
```

### PR DESCRIPTION
**Pull request by N5N3**
This PR recover the orignal behavior of `stride` for `Union{DenseArray,StridedReshapedArray,StridedReinterpretArray}`.
Besides, `stride(a::AbstractArray{<:Any, 0}, k)` now always return `1` if `k >= 1` and `strides` is defined.

Test added. (Close #44087)

### CODE REVIEW

**Comment by N5N3** [comment]
I have no idea. For me we can return any value if `k > ndims(A)`.
But since `Strided.jl` hardcoded a `stride` test.
I guess we'd better keep the return as before if possible?

File: base/abstractarray.jl [hunk 1]
```diff
@@ -546,7 +546,7 @@ julia> stride(A,3)
 function stride(A::AbstractArray, k::Integer)
     st = strides(A)
     k ≤ ndims(A) && return st[k]
-    return sum(st .* size(A))
+    return ndims(A) == 0 ? 1 : sum(st .* size(A))
```
Comment (JeffBezanson) [inline base/abstractarray.jl:549]: This seems like a very low-level function to be calling broadcast and reduce. Should probably be a loop.

Comment (mkitti) [reply base/abstractarray.jl:549]: I suggest restoring this function exactly:
```julia
function _stride(a, i)
    if i > ndims(a)
        return length(a)
    end
    s = 1
    for n = 1:(i-1)
        s *= size(a, n)
    end
    return s
end
```

https://github.com/JuliaLang/julia/blob/dd0c14ba1d0add2ce89524a26684a1194a83312c/base/reinterpretarray.jl#L162-L171

Comment (N5N3) [reply base/abstractarray.jl:549]: The orginal version don't thrown a error for `stride(a, -1)`.
I think it's meaningful the keep the consistency of error message. The easist way is follow the implement in general fallback.
(And they should have no performance difference)

**Comment by mkitti** [comment]
@JeffBezanson , @N5N3 . I'm confused. I get a distinct result with this new implementation.

Julia 1.7.2:
```julia
julia> stride(zeros(5,3,2),1)
1

julia> stride(zeros(5,3,2),2)
5

julia> stride(zeros(5,3,2),3)
15

julia> stride(zeros(5,3,2),4)
30
```

This function:

```julia
julia> function new_stride(A::AbstractArray, k::Integer)
           st = strides(A)
           k ≤ ndims(A) && return st[k]
           ndims(A) == 0 && return 1
           sz = size(A)
           s = st[1] * sz[1]
           for i in 2:ndims(A)
               s += st[i] * sz[i]
           end
           return s
       end
new_stride (generic function with 1 method)

julia> new_stride(zeros(5,3,2),1)
1

julia> new_stride(zeros(5,3,2),2)
5

julia> new_stride(zeros(5,3,2),3)
15

julia> new_stride(zeros(5,3,2),4)
50
```

Why are we summing?
```
julia> strides(A) .* size(A)
(5, 15, 30)

julia> strides(A) .* size(A) |> sum
50
```

**Comment by N5N3** [comment]
We only do sum for non-dense array. see https://github.com/JuliaLang/julia/pull/35929#discussion_r446208039 for more details.

**Comment by musm** [comment]
> Upon reflection here I still do have a slight preference for return sum(st .* size(A)) here simply because using length alone is a code smell in strides computations — it just looks wrong. Indeed, it makes it clearer that we need to call strides in this branch (like Tim noted above). But as a style thing, I'll leave the final decision to you.


I'm not sure I follow  the logic either for `k > ndims(A)`. Sounds like  the preference to sum is to avoid  using `length` ?

**Comment by N5N3** [comment]
I guess yes. Although it's correctness is fragile as stride could `< 0`.

**Pull request merged**